### PR TITLE
Endpoint router rule cache [ESD-117]

### PR DESCRIPTION
**Pull request by silverjam**
+ Use a rule cache to make processing of router rules more efficient

+ Previously the router would process rules one at a time as data arrived in order to find any matching rules.  Instead the list of rules can be pre-processed using the prefixes specified in the rule list to build a hash map that maps a rule prefix to a list of ports.  When data arrives, the first N bytes is used to look-up the list of ports that the data should be sent to

+ This results a drop in the latency required to process a packet as seen below and should also result in a drop in the max latency seen for packets that match many rules or don't match a rule until the very end of a rule set.

### CODE REVIEW

**Comment by silverjam** [comment]
Blow are plots showing the latency in packet processing for `endpoint_router`

## Average latency of master:

latency_plot previous

## Average latency of this branch

latency_plot

## Max latency of master

latency_max_plot previous

## Max latency of this branch

latency_max_plot

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.1.0-develop-2018120522-13-geb7b9b4b

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=800917)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018120522-13-geb7b9b4b&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018120522-13-geb7b9b4b&build_type=buildroot_pr&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.1.0-develop-2018120522-13-geb7b9b4b
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2018120522-13-geb7b9b4b)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2018120522-13-geb7b9b4b/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2018120522-13-geb7b9b4b

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.1.0-develop-2018120522-15-g9fb3fc3a

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=801096)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018120522-15-g9fb3fc3a&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018120522-15-g9fb3fc3a&build_type=buildroot_pr&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.1.0-develop-2018120522-15-g9fb3fc3a
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2018120522-15-g9fb3fc3a)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2018120522-15-g9fb3fc3a/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2018120522-15-g9fb3fc3a

File: package/endpoint_router/endpoint_router/src/endpoint_router.c [hunk 1]
```diff
@@ -399,47 +445,321 @@ static void loop_1s_metrics(pk_loop_t *loop, void *handle, int status, void *con
...
+  router_cfg_teardown(&router->router_cfg);
+
+  for (size_t rule_idx = 0; rule_idx < router->port_count; rule_idx++) {
+
+    rule_cache_t *rule_cache = &router->port_rule_cache[rule_idx];
+
+    if (rule_cache->cached_ports != NULL) {
+
+      for (size_t idx = 0; idx < rule_cache->rule_prefixes->count; idx++) {
+        if (rule_cache->cached_ports->endpoints != NULL) {
```
Comment (pmiettinen) [inline package/endpoint_router/endpoint_router/src/endpoint_router.c]: Is this correct? Missing `[idx]`?

File: package/endpoint_router/endpoint_router/src/endpoint_router_load.c [hunk 2]
```diff
@@ -19,6 +19,9 @@
 
 #include "endpoint_router_load.h"
 
+// Override-able for unit testing
```
Comment (pmiettinen) [inline package/endpoint_router/endpoint_router/src/endpoint_router_load.c]: c89 style preferred

File: package/endpoint_router/endpoint_router/src/endpoint_router.c (hunk 1 above)
Comment (silverjam) [reply package/endpoint_router/endpoint_router/src/endpoint_router.c]: fixed

File: package/endpoint_router/endpoint_router/src/endpoint_router_load.c (hunk 2 above)
Comment (silverjam) [reply package/endpoint_router/endpoint_router/src/endpoint_router_load.c]: fixed

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v2.1.0-develop-2018120522-19-gdb609217

These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.

### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=807474)

### gnss-analysis results
At least one run must complete for these links to have data.

#### passfail"
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018120522-19-gdb609217&build_type=buildroot_pr&display_type=table)

#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/metrics_preset=detailed&scenario=live-roof-650-townsend-post&summary_type=q50&groupby_key=firmware&firmware_versions=v2.1.0-develop-2018120522-19-gdb609217&build_type=buildroot_pr&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v2.1.0-develop-2018120522-19-gdb609217
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2018120522-19-gdb609217)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2018120522-19-gdb609217/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2018120522-19-gdb609217

**Pull request merged**